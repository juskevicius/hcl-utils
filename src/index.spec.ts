import fs from 'fs';
import { HclUtils } from './index';

describe('hcl utils,', () => {
  const hclUtils = new HclUtils();
  const tfFiles = ['arr.tf', 'comments.tf', 'eof.tf', 'math.tf', 'methods.tf', 'other.tf', 'strings.tf'];
  describe.each(tfFiles)('when running on file %s,', (fileName) => {
    const inputTfFile = fs.readFileSync(`./src/test-sample-data/${fileName}`).toString();
    let parsed: any;

    it('does parsing correctly', () => {
      parsed = hclUtils.parse(inputTfFile);
      expect(parsed).toMatchSnapshot();
    });

    it('builds back an identical file', () => {
      const outputTfFile = hclUtils.compose(parsed);
      expect(outputTfFile).toEqual(inputTfFile);
    });
  });
});
