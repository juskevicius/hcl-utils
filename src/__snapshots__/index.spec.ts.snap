// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hcl utils, when running on file arr.tf, does parsing correctly 1`] = `
Object {
  "_blank_line#27": "",
  "_formatting": Object {
    "withEqualsSign": Array [],
  },
  "resource \\"aws_security_group\\" \\"prod_sg\\"": Object {
    "_formatting": Object {
      "withEqualsSign": Array [],
    },
    "ingress": Object {
      "_formatting": Object {
        "withEqualsSign": Array [
          "cidr_blocks",
          "stuff",
          "otherStuff",
          "strazngeStuff",
          "moreStuff",
          "records",
          "records2",
        ],
      },
      "cidr_blocks": Array [
        "\\"0.0.0.0/0\\"",
      ],
      "moreStuff": Array [
        "\\"text\\"",
        "\\"moreText\\"",
      ],
      "otherStuff": Array [],
      "records": Array [
        "\\"\${data.dns_a_record_set.gce_production_1_nat.addrs}\\"",
        "\\"\${data.dns_a_record_set.gce_production_2_nat.addrs}\\"",
        "\\"\${data.dns_a_record_set.gce_production_3_nat.addrs}\\"",
      ],
      "records2": Array [
        "\\"\${
      distinct(
        concat(
          data.dns_a_record_set.gce_production_1_build_cache.addrs,
          data.dns_a_record_set.gce_production_2_build_cache.addrs
        )
      )
    }\\"",
      ],
      "strazngeStuff": Array [
        "data.dns_a_record_set.gce_production_1",
      ],
      "stuff": Array [
        "\\"asdf\\"",
        "\\"asdf\\"",
        "\\"asdf\\"",
        "\\"asdf\\"",
        "\\"asdf\\"",
        "\\"asdf\\"",
        "\\"fff\\"",
        "\\"bbb\\"",
        "\\"asdf\\"",
        "\\"asdf\\"",
        "\\"asdf\\"",
        "\\"asdf\\"",
      ],
    },
  },
}
`;

exports[`hcl utils, when running on file comments.tf, does parsing correctly 1`] = `
Object {
  "_blank_line#17": "",
  "_blank_line#21": "",
  "_comment#1": "# Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
  "_comment#13": "
// Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
  "_comment#18": "/*
Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
*/",
  "_formatting": Object {
    "withEqualsSign": Array [],
  },
  "output \\"some_val\\"": Object {
    "_formatting": Object {
      "withEqualsSign": Array [
        "value",
      ],
    },
    "value": "module.joujou.host",
  },
  "resource \\"aws_key_pair\\" \\"prod_ec2_ssh_key_pair\\"": Object {
    "_formatting": Object {
      "withEqualsSign": Array [],
    },
    "provisioner \\"local-exec\\"": Object {
      "_formatting": Object {
        "withEqualsSign": Array [
          "command",
        ],
      },
      "command": "\\"echo '\${tls_private_key.prod_tls_private_key.private_key_pem}' > ./myKey.pem\\"",
    },
    "provisioner \\"local-exec\\"#6": Object {
      "_formatting": Object {
        "withEqualsSign": Array [
          "command",
        ],
      },
      "command": "\\"chmod 400 ./myKey.pem\\"",
    },
    "provisioner \\"local-exec\\"#9": Object {
      "_formatting": Object {
        "withEqualsSign": Array [
          "command",
        ],
      },
      "command": "\\"echo 'aaa'\\"",
    },
  },
  "variable \\"project\\"": Object {
    "_formatting": Object {
      "withEqualsSign": Array [
        "default",
      ],
    },
    "default": "\\"eco-hjg-663\\"",
  },
}
`;

exports[`hcl utils, when running on file eof.tf, does parsing correctly 1`] = `
Object {
  "_blank_line#12": "",
  "_formatting": Object {
    "withEqualsSign": Array [],
  },
  "resource \\"null_resource\\" \\"askdj\\"": Object {
    "_blank_line#5": "",
    "_formatting": Object {
      "withEqualsSign": Array [],
    },
    "provisioner \\"local-exec\\"": Object {
      "_formatting": Object {
        "withEqualsSign": Array [
          "command",
        ],
      },
      "command": "<<EOF
exec \${path.module}/../bin/heroku-stop-me \\\\
  --repo=travis-ci/whereami \\\\
  --app=\${heroku_app.whereami.id} \\\\
  --ps-scale=\${join(\\",\\", var.whereami_scale)} \\\\
  --deploy-version=\${var.whereami_version}
EOF",
    },
    "website": Object {
      "_blank_line#9": "",
      "_formatting": Object {
        "withEqualsSign": Array [
          "index_document",
          "error_document",
          "routing_rules",
        ],
      },
      "error_document": "\\"error.html\\"",
      "index_document": "\\"index.html\\"",
      "routing_rules": "<<EOF
[{
    \\"Condition\\": {
        \\"KeyPrefixEquals\\": \\"docs/\\"
    },
    \\"Redirect\\": {
        \\"ReplaceKeyPrefixWith\\": \\"documents/\\"
    }
}]
EOF",
    },
  },
}
`;

exports[`hcl utils, when running on file math.tf, does parsing correctly 1`] = `
Object {
  "_blank_line#6": "",
  "_formatting": Object {
    "withEqualsSign": Array [],
  },
  "resource \\"aws_key_pair\\" \\"prod_ec2_ssh_key_pair\\"": Object {
    "_formatting": Object {
      "withEqualsSign": Array [
        "math",
        "math2",
        "math3",
        "math4",
      ],
    },
    "math": "merge(var.count) + var.id",
    "math2": "var.jjjj - 17",
    "math3": "var.lll * 16",
    "math4": "var.kkkk / 15",
  },
}
`;

exports[`hcl utils, when running on file methods.tf, does parsing correctly 1`] = `
Object {
  "_blank_line#11": "",
  "_blank_line#8": "",
  "_formatting": Object {
    "withEqualsSign": Array [],
  },
  "resource \\"null_resource\\" \\"askdj\\"": Object {
    "_formatting": Object {
      "withEqualsSign": Array [],
    },
    "triggers": Object {
      "_formatting": Object {
        "withEqualsSign": Array [
          "some_stuff",
          "config_signature",
          "scale",
        ],
      },
      "config_signature": "sha256(join(\\",\\", values(jjjjj.asdjhf.ll_jjj.0)))",
      "scale": "join(\\",\\", var.some_var)",
      "some_stuff": "slice(var.ss, 0)",
    },
  },
  "variable \\"project\\"": Object {
    "_formatting": Object {
      "withEqualsSign": Array [
        "config_signature",
      ],
    },
    "config_signature": "sha256(join(\\",\\", values(super_app.my_vars.0)))",
  },
}
`;

exports[`hcl utils, when running on file other.tf, does parsing correctly 1`] = `
Object {
  "_blank_line#2": "",
  "_blank_line#26": "",
  "_blank_line#3": "",
  "_formatting": Object {
    "withEqualsSign": Array [],
  },
  "resource \\"aws_key_pair\\" \\"prod_ec2_ssh_key_pair\\"": Object {
    "_blank_line#7": "",
    "_formatting": Object {
      "withEqualsSign": Array [
        "num",
        "bool",
      ],
    },
    "bool": "true",
    "num": "5",
    "provisioner \\"local-exec\\"": Object {
      "_formatting": Object {
        "withEqualsSign": Array [
          "command",
        ],
      },
      "command": "\\"echo '\${tls_private_key.prod_tls_private_key.private_key_pem}' > ./myKey.pem\\"",
    },
    "provisioner \\"local-exec\\"#11": Object {
      "_formatting": Object {
        "withEqualsSign": Array [
          "command",
        ],
      },
      "command": "\\"chmod 400 ./myKey.pem\\"",
    },
    "provisioner \\"local-exec\\"#14": Object {
      "_formatting": Object {
        "withEqualsSign": Array [
          "command",
        ],
      },
      "command": "\\"echo 'aaa'\\"",
    },
    "provisioner \\"local-exec\\"#17": Object {
      "_formatting": Object {
        "withEqualsSign": Array [],
      },
      "deep": Object {
        "_formatting": Object {
          "withEqualsSign": Array [],
        },
        "nested": Object {
          "_formatting": Object {
            "withEqualsSign": Array [],
          },
          "property": Object {
            "_formatting": Object {
              "withEqualsSign": Array [
                "equals",
              ],
            },
            "equals": "\\"this\\"",
          },
        },
      },
    },
  },
  "variable \\"github_users\\"": Object {
    "_formatting": Object {
      "withEqualsSign": Array [],
    },
  },
}
`;

exports[`hcl utils, when running on file strings.tf, does parsing correctly 1`] = `
Object {
  "_blank_line#17": "",
  "_blank_line#23": "",
  "_blank_line#28": "",
  "_blank_line#38": "",
  "_blank_line#7": "",
  "_formatting": Object {
    "withEqualsSign": Array [],
  },
  "output \\"some_val\\"": Object {
    "ENVIRONMENT": "var.environment",
    "_formatting": Object {
      "withEqualsSign": Array [
        "value",
        "ENVIRONMENT",
      ],
    },
    "value": "module.joujou.host",
  },
  "resource \\"aws_key_pair\\" \\"prod_ec2_ssh_key_pair\\"": Object {
    "_formatting": Object {
      "withEqualsSign": Array [
        "key_name",
        "some_val",
        "public_key",
      ],
    },
    "key_name": "\\"myKey\\"",
    "public_key": "tls_private_key.prod_tls_private_key.public_key_openssh",
    "some_val": "\\"http://\${local.thing}/path?param\\"",
  },
  "resource \\"null_resource\\" \\"askdj\\"": Object {
    "_formatting": Object {
      "withEqualsSign": Array [],
    },
    "triggers": Object {
      "_formatting": Object {
        "withEqualsSign": Array [
          "heroku_id",
          "version",
        ],
      },
      "heroku_id": "super_app.some_name.id",
      "version": "var.some_version",
    },
  },
  "terraform": Object {
    "_formatting": Object {
      "withEqualsSign": Array [],
    },
    "backend \\"s3\\"": Object {
      "_formatting": Object {
        "withEqualsSign": Array [
          "bucket",
          "key",
          "region",
          "encrypt",
          "dynamodb_table",
        ],
      },
      "bucket": "\\"random-terraform-state\\"",
      "dynamodb_table": "random-terraform.*.state",
      "encrypt": "\\"true\\"",
      "key": "\\"terraform-config/build-caching-production-1.tfstate\\"",
      "region": "\\"us-east-1\\"",
    },
  },
  "variable \\"project\\"": Object {
    "SOME_IPS": "\\"\${
    join(\\",\\", data.production_1_nat.addrs)
    },\${
    join(\\",\\", data.production_2_nat.addrs)
    },\${
    join(\\",\\", data.production_3_nat.addrs)
  }\\"",
    "_formatting": Object {
      "withEqualsSign": Array [
        "default",
        "SOME_IPS",
      ],
    },
    "default": "\\"eco-hjg-663\\"",
  },
}
`;
